### PR TITLE
[WIP] Implement Potion Effects

### PR DESCRIPTION
**Pull request by walker27460**
# Implement Potion Effects

## Status

- [ ] Ready 
- [x] Development
- [ ] Hold

## Description

Implemented Potion Effects.

## Checklist

- [x] Ran `cargo fmt`, `cargo clippy`, `cargo build --release` and `cargo test` and fixed any generated errors!
- [x] Removed unnecessary commented out code
- [ ] Used specific traces (if you trace actions please specify the cause i.e. the player)

### CODE REVIEW

File: feather/common/src/entities/player.rs [hunk 1]
```diff
@@ -11,6 +11,7 @@ pub fn build_default(builder: &mut EntityBuilder) {
     builder
         .add(Player)
         .add(CreativeFlying(false))
+        .add(PotionEffects(vec![]))
```
Comment (OxKitsune) [inline feather/common/src/entities/player.rs:14]: Do we need this? Is it not better to only add the component when the player actually has potion effects?

**Comment by OxKitsune** [comment]
Maybe some utility methods for adding potion effects to players need to be added? Just like the particle spawning methods, cause you could just add the components to the player. But I think having a util method would be good.

File: feather/common/src/entities/player.rs (hunk 1 above)
Comment (walker27460) [reply feather/common/src/entities/player.rs:14]: maybe that will be better?
if so then I'm gonna remove that line

File: quill/common/src/components.rs [hunk 2]
```diff
@@ -106,20 +107,73 @@ pub struct CreativeFlying(pub bool);
...
+impl_effect!(Saturation);
+impl_effect!(Glowing);
+impl_effect!(Levitation);
+impl_effect!(Luck);
+impl_effect!(BadLuck);
+impl_effect!(SlowFalling);
+impl_effect!(ConduitPower);
+impl_effect!(DolphinsGrace);
+impl_effect!(BadOmen);
+impl_effect!(HeroOfTheVillage);
```
Comment (amberkowalski) [inline quill/common/src/components.rs:176]: All of these should have the `Effect` suffix.

File: quill/common/src/components.rs [hunk 3]
```diff
@@ -106,20 +107,73 @@ pub struct CreativeFlying(pub bool);
...
-    particle: bool,
-    ambient: bool, // given from beacon or not.
-    icon: bool,    // show effect icon.
+#[derive(Copy, Clone, Debug, Hash, Serialize, PartialEq, Eq, Deserialize)]
+pub struct PotionApplication {
+    pub amplifier: u8,
+    pub duration: u32,
+    pub particle: bool,
+    pub ambient: bool, // given from beacon or not.
+    pub icon: bool,    // show effect icon.
```
Comment (amberkowalski) [inline quill/common/src/components.rs:116]: Switch the comments to doc comments on the members, consider adding one for `particle` and `duration`.

File: quill/common/src/components.rs [hunk 4]
```diff
@@ -106,20 +107,73 @@ pub struct CreativeFlying(pub bool);
...
+}
+
+macro_rules! impl_effect {
+    ($ident:ident) => {
+        #[derive(Serialize, Deserialize, Eq, PartialEq, Hash)]
+        pub struct $ident(BTreeSet<PotionApplication>);
+        impl $ident {}
+        bincode_component_impl!($ident);
+    };
+}
```
Comment (amberkowalski) [inline quill/common/src/components.rs:143]: I think we might have somewhere we try to consolidate macros? Not sure.

File: quill/common/src/components.rs (hunk 3 above)
Comment (amberkowalski) [inline quill/common/src/components.rs:116]: We could probably move effects into their own module.

**Review by amberkowalski — COMMENTED** [review]
Just a few things I see, I know its not done yet.

File: quill/common/src/components.rs (hunk 3 above)
Comment (walker27460) [reply quill/common/src/components.rs]: the effect enum was originally in libcraft but I made a transition to components.

**Comment by amberkowalski** [comment]
Hi there, it appears your PR has been inactive for a while. Are you planning to continue it? Please let me know so I can either pick it up or wait for more progress.

**Comment by walker27460** [comment]
yeah im try to continue but no progress

**Comment by Defman** [comment]
Bump

**Pull request closed without merging**